### PR TITLE
Firestore indexes bug fixes

### LINKED ISSUE #1075: firestore:indexed Error: An unexpected error has occurred.
I think the new exemption field index cause a problem with tools exporting command

The problem arise when there was an exemption that was disable every indexing

like this

image

Will cause error like this

image

### LINKED ISSUE #1080: Need to call several times: firebase deploy --only firestore:indexes  to get all indexes created
### Version info
6.2.2

### Platform Information
OS X 10.14.2

### Steps to reproduce
delete all the firestore indexes manually through Firebase UI
then call:
firebase deploy --only firestore:indexes
with a file containing lot of indexes (mine's got 64 indexes)

### Expected behavior
create all the indexes (64) described inside my file firestore.indexes.json

### Actual behavior
It creates only the first 7 indexes. I need to call several times the same command:
"firebase deploy --only firestore:indexes"
to create some new indexes on each command
second time added 8 indexes...
(can't know how make it worked with the  "--debug flag")

If this is normal behaviour, maybe I missed the explanation inside the doc?

And happy new year ;-) !

### LINKED ISSUE #1081: no indexes being built with 6.2.2
i've never had an issue creating indexes with the CLI before version 6.2.2.  now, when i deploy the following as a test:

```
{
  "indexes": [
    {
      "collectionId": "replies",
      "fields": [
        { "fieldPath": "topicId", "mode": "ASCENDING" },
        { "fieldPath": "createdAt", "mode": "DESCENDING" }
      ]
    }
  ]
}

```

i get the standard success message of:

```
i  deploying firestore
i  firestore: reading indexes from firestore.indexes.json...
i  firestore: your indexes indexes are specified in the v1beta1 API format. Please upgrade to the new index API format by running firebase firestore:indexes again and saving the result.
✔  firestore: deployed indexes in firestore.indexes.json successfully
✔  Deploy complete!
```

except the index was never created.  when i then deploy a v1beta1 version:

```
{
  "indexes": [
    {
      "collectionGroup": "replies",
      "queryScope": "COLLECTION",
      "fields": [
        {
          "fieldPath": "topicId",
          "order": "ASCENDING"
        },
        {
          "fieldPath": "createdAt",
          "order": "DESCENDING"
        }
      ]
    }
  ],
  "fieldOverrides": []
}
```

i get the same success message (w/o the warning), but the index is again not created.  

### Version info
6.2.2
### Platform Information
OS X
### Steps to reproduce
firebase deploy --only firestore:indexes

### PR DESCRIPTION
**Pull request by samtstern**
Changes:

  * Fixes #1080 and Fixes #1081 
  * Fixes #1075 
  * Fixes empty state for pretty print

### CODE REVIEW

File: src/firestore/indexes.ts [hunk 1]
```diff
@@ -59,16 +59,17 @@ export class FirestoreIndexes {
       );
     }
 
-    indexesToDeploy.forEach(async (index) => {
+    const indexPromises: Array<Promise<any>> = [];
+    for (const index of indexesToDeploy) {
```
Comment (samtstern) [inline src/firestore/indexes.ts]: TIL: If you do `forEach` with an `async` function inside it you just generate a big fat race condition.

File: src/firestore/indexes-api.ts [hunk 2]
```diff
@@ -65,5 +65,5 @@ export interface Field {
  */
 export interface IndexConfig {
   ancestorField?: string;
-  indexes: Index[];
```
Comment (samtstern) [inline src/firestore/indexes-api.ts]: I thought the API would return `indexes: []` in this case but it's really `indexes: undefined`.  Changed the type here, fixed all call sites, added a test to make sure we consider this equivalent.

File: src/firestore/indexes.ts (hunk 1 above)
Comment (bkendall) [reply src/firestore/indexes.ts]: `forEach` is sync, so you could `await` your `async` function in `forEach` (but this is just silly):

```typescript
indexesToDeploy.forEach(await (async (index) => { /* do stuff */ }));
```

The alternative to this `forEach`-populating-an-array is to use `map` and return the promise at the end, then `Promise.all` them (though, it looks like you have conditionals here, so maybe `forEach` is correct!)

File: src/firestore/indexes.ts [hunk 3]
```diff
@@ -563,11 +577,16 @@ export class FirestoreIndexes {
...
-      const firstField = index.fields[0];
-      const mode = firstField.order || firstField.arrayConfig;
-      result += " (" + mode + ")";
-    });
+    const fieldIndexes = field.indexConfig.indexes || [];
+    if (fieldIndexes.length > 0) {
+      fieldIndexes.forEach((index) => {
+        const firstField = index.fields[0];
+        const mode = firstField.order || firstField.arrayConfig;
+        result += " (" + mode + ")";
```
Comment (bkendall) [inline src/firestore/indexes.ts]: prefer template strings:

```suggestion
        result += ` (${mode})`;
```

**Pull request merged**